R/clustering.R

```diff
@@ -476,3 +480,42 @@ RunModularityClustering <- function(
   )
   return(clusters)
 }
+
+# Set up Leiden clustering algorithm
+# @name Leiden State Matrix
+# @rdname leiden
+#
+# @title Run Leiden clustering algorithm
+#
+# @description Implements the Leiden clustering algorithm in R using reticulate to run the Python version. Requires the python "leidenalg" and "igraph" modules to be installed. Returns a vector of partition indices.
+#
+# @param object A seurat class object.
+# @param ... Parameters to pass to the Python leidenalg function.
+#
+# @keywords graph network igraph mvtnorm simulation
+# @import reticulate
```

Comment:
I think I've done this correctly but I'm not sure how roxygen handles internal functions. I'm unsure if importing only these specific functions from reticulate are sufficient for the Python interface to work. This may need to be tested more extensively.

Reply:
`@import` makes the entire namespace available for the function being documented, `@importFrom` makes only specific functions available. In Python terms, this is akin to `from package import *` versus `from package import function1, function2`; this doesn't mean that the functions imported don't see everything from their original package, just that we don't. Therefore, just importing the functions you call from reticulate doesn't mean that the Python objects you create in R don't have access to the full reticulate package. Those objects still have access to all necessary code to enable them to function.

We prefer the use of `@importFrom` for declaring dependencies for two reasons: being explicit and being lazy. Using `@importFrom` over `@import` requires that we declare each and every function from a dependency package, which means we can keep track of where our code comes from (`base` and Seurat-defined functions aren't declared, everything else is) and which version of each function we're running. It also means we can reduce namespace clash within Seurat, as several packages name functions the same (eg. `Matrix::colSums` and `base::colSums`).

We could require that everyone uses the `package::function` construct (analogous to Python's `module.function` construct), but then we lose Roxygen's automatic dependency declaration in the NAMESPACE file, and it's very easy to turn a NAMESPACE into a list of imports for the DESCRIPTION (all stuff required for CRAN). Since we're too lazy to comb through Seurat and keep track of every unique dependency, we let Roxygen do it for us with the `@importFrom` tags.

Hopefully this makes sense and assures your concerns about use of this construct to declare dependencies.